[forge] add PFN support

rustielin opened the pull request:
### Description

Support adding PFNs to Forge. For now, PFNs are added in the test logic (during runtime), rather than specified in the test config.

The bulk of the changes introduced in this PR are around the new `fullnode` module in the Forge k8s backend, that handles creating k8s resources to define a PFN. In particular, the k8s resources are created directly using the kube-rs crate rather than going through an external tool like helm, which allows for more flexibility. The following are created:
* a `ConfigMap` with the NodeConfig
* a `StatefulSet` to run the actual PFN. In the spec, use an embedded `PersistentVolumeClaim` to define the storage volume requirements of the PFN.
* a `Service` to allow traffic to the PFN

For simplicity, much of the `StatefulSet` and `PersistentVolumeClaim` is inherited from the validators, specifically validator-0 (which is always guaranteed to be defined). The genesis and waypoint can be mounted from the validator secrets from genesis, and the size of the storage volume should be the same as that of the validator. 

Changes to Forge interface:
* add `Swarm::get_default_pfn_node_config`, which is useful as a base for user-specified PFN NodeConfigs. Test writers would use the default node config and override fields in the NodeConfig. This will be used to write indexer gRPC performance tests in Forge, which is a TODO!
* make `Swarm::add_full_node` async

Misc Forge changes:
* wait for nodes to come up using a fixed delay retry, rather than exponential, to make it faster. The spin up time is mostly predictable too, so it makes sense to use fixed
* new `mock_kube_apis` module in k8s backend that is used in unit testing k8s API in-memory

Misc changes:
* move NodeConfig merging logic from `aptos-node` to `config`
* remove unused`KUBERNETES_POD_NAME` env var in `terraform/helm/aptos-node/templates/validator.yaml`
* minor change to `forge.py` in calculating image repo


### Test Plan

New unit tests for new PFN creation logic.

Update the existing `forge_setup_test` to add a PFN and send load to it. Run this using the local Forge runner mode, targeting the k8s backend:

```
$ FORGE_RUNNER_MODE=local FORGE_TEST_SUITE=setup_test ./testsuite/testrun forge.py test 
...
2023-06-07T18:08:45.689163Z [main] INFO testsuite/forge/src/backend/k8s/fullnode.rs:423 Wrote fullnode k8s specs to path: TempDir { path: "/var/folders/0r/0cqndyk91kb6s_xv23gdvsdw0000gn/T/.tmpHWkMJ5" }
2023-06-07T18:08:45.774318Z [main] INFO testsuite/forge/src/backend/k8s/fullnode.rs:429 Created fullnode fullnode-78d29195 StatefulSet: StatefulSet { metadata: ObjectMeta { annotations: None, cluster_name: None, creation_timestamp: None, deletion_grace_period_seconds: None, deletion_timestamp: None, finalizers: None, generate_name: None, generation: None, labels: Some({"app.kubernetes.io/instance": "fullnode-78d29195", "app.kubernetes.io/name": "fullnode", "app.kubernetes.io/part-of": "forge-pfn"}), managed_fields: None, name: Some("fullnode-78d29195"), namespace: None, owner_references: None, resource_version: None, self_link: None, uid: None }, spec: Some(StatefulSetSpec { min_ready_seconds: None, pod_management_policy: Some("Parallel"), replicas: Some(1), revision_history_limit: Some(10), selector: LabelSelector { match_expressions: None, match_labels: Some({"app.kubernetes.io/instance": "fullnode-78d29195", "app.kubernetes.io/name": "fullnode", "app.kubernetes.io/part-of": "forge-pfn"}) }, service_name: "fullnode-78d29195", template: PodTemplateSpec { metadata: Some(ObjectMeta { annotations: None, cluster_name: None, creation_timestamp: None, deletion_grace_period_seconds: None, deletion_timestamp: None, finalizers: None, generate_name: None, generation: None, labels: Some({"app.kubernetes.io/instance": "fullnode-78d29195", "app.kubernetes.io/name": "fullnode", "app.kubernetes.io/part-of": "forge-pfn"}), managed_fields: None, name: None, namespace: None, owner_references: None, resource_version: None, self_link: None, uid: None }), spec: Some(PodSpec { active_deadline_seconds: None, affinity: Some(Affinity { node_affinity: Some(NodeAffinity { preferred_during_scheduling_ignored_during_execution: None, required_during_scheduling_ignored_during_execution: Some(NodeSelector { node_selector_terms: [NodeSelectorTerm { match_expressions: Some([NodeSelectorRequirement { key: "cloud.google.com/machine-family", operator: "In", values: Some(["t2d"]) }]), match_fields: None }] }) }), pod_affinity: None, pod_anti_affinity: Some(PodAntiAffinity { preferred_during_scheduling_ignored_during_execution: None, required_during_scheduling_ignored_during_execution: Some([PodAffinityTerm { label_selector: Some(LabelSelector { match_expressions: Some([LabelSelectorRequirement { key: "app.kubernetes.io/part-of", operator: "In", values: Some(["aptos-node"]) }]), match_labels: None }), namespace_selector: None, namespaces: None, topology_key: "kubernetes.io/hostname" }]) }) }), automount_service_account_token: None, containers: [Container { args: None, command: Some(["/usr/local/bin/aptos-node", "-f", "/opt/aptos/etc/fullnode.yaml"]), env: Some([EnvVar { name: "RUST_LOG", value: Some("debug,hyper=off"), value_from: None }, EnvVar { name: "APTOS_FORCE_ENABLE_TELEMETRY", value: Some("true"), value_from: None }, EnvVar { name: "KUBERNETES_NAMESPACE", value: None, value_from: Some(EnvVarSource { config_map_key_ref: None, field_ref: Some(ObjectFieldSelector { api_version: Some("v1"), field_path: "metadata.namespace" }), resource_field_ref: None, secret_key_ref: None }) }, EnvVar { name: "RUST_BACKTRACE", value: Some("0"), value_from: None }]), env_from: None, image: Some("us-west1-docker.pkg.dev/aptos-global/aptos-internal/validator-testing:09359aa961e909699939783bd438e967ca381bdb"), image_pull_policy: Some("IfNotPresent"), lifecycle: None, liveness_probe: None, name: "validator", ports: Some([ContainerPort { container_port: 6180, host_ip: None, host_port: None, name: None, protocol: Some("TCP") }, ContainerPort { container_port: 6181, host_ip: None, host_port: None, name: None, protocol: Some("TCP") }, ContainerPort { container_port: 6182, host_ip: None, host_port: None, name: None, protocol: Some("TCP") }, ContainerPort { container_port: 8080, host_ip: None, host_port: None, name: Some("api"), protocol: Some("TCP") }, ContainerPort { container_port: 9101, host_ip: None, host_port: None, name: Some("metrics"), protocol: Some("TCP") }]), readiness_probe: None, resources: Some(ResourceRequirements { limits: Some({"cpu": Quantity("30"), "memory": Quantity("60Gi")}), requests: Some({"cpu": Quantity("30"), "memory": Quantity("60Gi")}) }), security_context: Some(SecurityContext { allow_privilege_escalation: None, capabilities: None, privileged: Some(true), proc_mount: None, read_only_root_filesystem: None, run_as_group: None, run_as_non_root: None, run_as_user: Some(0), se_linux_options: None, seccomp_profile: None, windows_options: None }), startup_probe: None, stdin: None, stdin_once: None, termination_message_path: Some("/dev/termination-log"), termination_message_policy: Some("File"), tty: None, volume_devices: None, volume_mounts: Some([VolumeMount { mount_path: "/opt/aptos/etc", mount_propagation: None, name: "aptos-config", read_only: None, sub_path: None, sub_path_expr: None }, VolumeMount { mount_path: "/opt/aptos/data", mount_propagation: None, name: "aptos-data", read_only: None, sub_path: None, sub_path_expr: None }, VolumeMount { mount_path: "/opt/aptos/genesis", mount_propagation: None, name: "genesis-config", read_only: None, sub_path: None, sub_path_expr: None }]), working_dir: None }], dns_config: None, dns_policy: Some("ClusterFirst"), enable_service_links: None, ephemeral_containers: None, host_aliases: None, host_ipc: None, host_network: None, host_pid: None, hostname: None, image_pull_secrets: None, init_containers: None, node_name: None, node_selector: None, overhead: None, preemption_policy: None, priority: None, priority_class_name: None, readiness_gates: None, restart_policy: Some("Always"), runtime_class_name: None, scheduler_name: Some("default-scheduler"), security_context: Some(PodSecurityContext { fs_group: Some(0), fs_group_change_policy: None, run_as_group: Some(0), run_as_non_root: None, run_as_user: Some(0), se_linux_options: None, seccomp_profile: None, supplemental_groups: None, sysctls: None, windows_options: None }), service_account: Some("aptos-node-validator"), service_account_name: Some("aptos-node-validator"), set_hostname_as_fqdn: None, share_process_namespace: None, subdomain: None, termination_grace_period_seconds: Some(0), tolerations: Some([Toleration { effect: Some("NoExecute"), key: Some("aptos.org/nodepool"), operator: None, toleration_seconds: None, value: Some("validators") }]), topology_spread_constraints: None, volumes: Some([Volume { aws_elastic_block_store: None, azure_disk: None, azure_file: None, cephfs: None, cinder: None, config_map: None, csi: None, downward_api: None, empty_dir: None, ephemeral: None, fc: None, flex_volume: None, flocker: None, gce_persistent_disk: None, git_repo: None, glusterfs: None, host_path: None, iscsi: None, name: "genesis-config", nfs: None, persistent_volume_claim: None, photon_persistent_disk: None, portworx_volume: None, projected: None, quobyte: None, rbd: None, scale_io: None, secret: Some(SecretVolumeSource { default_mode: None, items: None, optional: None, secret_name: Some("aptos-node-0-genesis-eforge116") }), storageos: None, vsphere_volume: None }, Volume { aws_elastic_block_store: None, azure_disk: None, azure_file: None, cephfs: None, cinder: None, config_map: Some(ConfigMapVolumeSource { default_mode: None, items: None, name: Some("fullnode-78d29195-config"), optional: None }), csi: None, downward_api: None, empty_dir: None, ephemeral: None, fc: None, flex_volume: None, flocker: None, gce_persistent_disk: None, git_repo: None, glusterfs: None, host_path: None, iscsi: None, name: "aptos-config", nfs: None, persistent_volume_claim: None, photon_persistent_disk: None, portworx_volume: None, projected: None, quobyte: None, rbd: None, scale_io: None, secret: None, storageos: None, vsphere_volume: None }]) }) }, update_strategy: Some(StatefulSetUpdateStrategy { rolling_update: Some(RollingUpdateStatefulSetStrategy { partition: Some(0) }), type_: Some("RollingUpdate") }), volume_claim_templates: Some([PersistentVolumeClaim { metadata: ObjectMeta { annotations: None, cluster_name: None, creation_timestamp: None, deletion_grace_period_seconds: None, deletion_timestamp: None, finalizers: None, generate_name: None, generation: None, labels: None, managed_fields: None, name: Some("aptos-data"), namespace: None, owner_references: None, resource_version: None, self_link: None, uid: None }, spec: Some(PersistentVolumeClaimSpec { access_modes: Some(["ReadWriteOnce"]), data_source: None, data_source_ref: None, resources: Some(ResourceRequirements { limits: None, requests: Some({"storage": Quantity("2Ti")}) }), selector: None, storage_class_name: None, volume_mode: None, volume_name: None }), status: None }]) }), status: None }
2023-06-07T18:08:45.855387Z [main] INFO testsuite/forge/src/backend/k8s/fullnode.rs:437 Created fullnode fullnode-78d29195 Service: Service { metadata: ObjectMeta { annotations: None, cluster_name: None, creation_timestamp: None, deletion_grace_period_seconds: None, deletion_timestamp: None, finalizers: None, generate_name: None, generation: None, labels: None, managed_fields: None, name: Some("fullnode-78d29195"), namespace: None, owner_references: None, resource_version: None, self_link: None, uid: None }, spec: Some(ServiceSpec { allocate_load_balancer_node_ports: None, cluster_ip: None, cluster_ips: None, external_ips: None, external_name: None, external_traffic_policy: None, health_check_node_port: None, internal_traffic_policy: None, ip_families: None, ip_family_policy: None, load_balancer_class: None, load_balancer_ip: None, load_balancer_source_ranges: None, ports: Some([ServicePort { app_protocol: None, name: None, node_port: None, port: 8080, protocol: None, target_port: None }]), publish_not_ready_addresses: None, selector: Some({"app.kubernetes.io/instance": "fullnode-78d29195", "app.kubernetes.io/name": "fullnode", "app.kubernetes.io/part-of": "forge-pfn"}), session_affinity: None, session_affinity_config: None, type_: None }), status: None }
2023-06-07T18:08:46.258608Z [main] INFO testsuite/forge/src/backend/k8s/stateful_set.rs:107 StatefulSet fullnode-78d29195 has 0/1 replicas
2023-06-07T18:08:46.332022Z [main] INFO testsuite/forge/src/backend/k8s/stateful_set.rs:157 Pod fullnode-78d29195-0 at phase Pending
2023-06-07T18:08:56.393898Z [main] INFO testsuite/forge/src/backend/k8s/stateful_set.rs:107 StatefulSet fullnode-78d29195 has 0/1 replicas
2023-06-07T18:08:56.451070Z [main] INFO testsuite/forge/src/backend/k8s/stateful_set.rs:145 Waiting for pod fullnode-78d29195-0
06-07T18:09:36.948766Z [main] INFO testsuite/forge/src/backend/k8s/stateful_set.rs:101 StatefulSet fullnode-78d29195 has scaled to 1
2023-06-07T18:09:37.952620Z [main] INFO testsuite/forge/src/backend/k8s/node.rs:91 Port-forward started for fullnode-78d29195 @ 127.0.0.1 from 59494 --> 8080
2023-06-07T18:09:38.121904Z [main] INFO crates/transaction-emitter-lib/src/emitter/mod.rs:274  Transaction emitter target mempool backlog is 40000
2023-06-07T18:09:38.121978Z [main] INFO crates/transaction-emitter-lib/src/emitter/mod.rs:279  Will use 1 clients and 2000 workers per client
2023-06-07T18:09:38.121996Z [main] INFO crates/transaction-emitter-lib/src/emitter/mod.rs:527 Will use 2000 workers per endpoint for a total of 2000 endpoint clients and 2000 accounts
2023-06-07T18:09:38.122082Z [main] INFO crates/transaction-emitter-lib/src/emitter/mod.rs:545 AccountMinter Seed (can be passed in to reuse accounts): [60, 246, 157, 90, 137, 200, 107, 204, 60, 175, 210, 227, 97, 217, 141, 17, 102, 77, 164, 94, 117, 98, 44, 241, 168, 6, 164, 76, 199, 187, 109, 90]
2023-06-07T18:09:38.122143Z [main] INFO crates/transaction-emitter-lib/src/emitter/mod.rs:556 Using reliable/retriable init transaction executor with 18 retries, every 10s
2023-06-07T18:09:38.122186Z [main] INFO crates/transaction-emitter-lib/src/emitter/account_minter.rs:81 Account creation plan created for 2000 accounts with 100004200500000 balance each.
2023-06-07T18:09:38.122223Z [main] INFO crates/transaction-emitter-lib/src/emitter/account_minter.rs:85     because of expecting 1000000000 txns and 2000000 gas at 100 gas price for each 
2023-06-07T18:09:38.122236Z [main] INFO crates/transaction-emitter-lib/src/emitter/account_minter.rs:106     through 40 seed accounts with 5000312025000000 each, each to fund 50 accounts
2023-06-07T18:09:38.276709Z [main] INFO crates/transaction-emitter-lib/src/emitter/account_minter.rs:127 Source account 000000000000000000000000000000000000000000000000000000000a550c18 current balance is 18446744073709551615, needed 200012483000000000 coins, or 1.084% of its balance
2023-06-07T18:09:38.276834Z [main] INFO crates/transaction-emitter-lib/src/emitter/account_minter.rs:284 Creating and funding seeds accounts
2023-06-07T18:09:43.392227Z [main] INFO crates/transaction-emitter-lib/src/emitter/account_minter.rs:191 Completed creating 40 seed accounts in 5s, each with 5000312025000000 coins, request stats: success 40, failed submit [0], failed wait [0]
2023-06-07T18:09:43.392289Z [main] INFO crates/transaction-emitter-lib/src/emitter/account_minter.rs:198 Creating additional 2000 accounts with 100004200500000 coins each
2023-06-07T18:09:49.981850Z [main] INFO crates/transaction-emitter-lib/src/emitter/account_minter.rs:250 Successfully completed creating 2000 accounts in 6s, request stats: success 2000, failed submit [0], failed wait [0]
2023-06-07T18:09:49.983902Z [main] INFO crates/transaction-emitter-lib/src/emitter/mod.rs:603 Checking account sequence and counting latency for 2000 out of 2000 total_workers
2023-06-07T18:09:50.013653Z [main] INFO crates/transaction-emitter-lib/src/emitter/mod.rs:637 Tx emitter workers started
2023-06-07T18:09:50.013763Z [main] INFO crates/transaction-emitter-lib/src/emitter/mod.rs:686 Starting emitting txns for 10 secs in 1 phases
2023-06-07T18:10:00.014902Z [main] INFO crates/transaction-emitter-lib/src/emitter/mod.rs:705 Ran for 10 secs, stopping job...

```

The resources look like
```
$ kgpo  
NAME                                    READY   STATUS              RESTARTS   AGE
aptos-node-0-fullnode-eforge116-0       1/1     Running             0          106s
aptos-node-0-validator-0                1/1     Running             0          3m23s
fullnode-78d29195-0                     0/1     ContainerCreating   0          44s
genesis-aptos-genesis-eforge116-xhcl6   0/1     Completed           0          3m41s

$ kg svc
NAME                     TYPE        CLUSTER-IP    EXTERNAL-IP   PORT(S)                                                       AGE
aptos-node-0-fullnode    NodePort    10.64.8.135   <none>        6182:30680/TCP,9101:30136/TCP,8080:32103/TCP                  3m41s
aptos-node-0-validator   NodePort    10.64.2.199   <none>        6180:31868/TCP,6181:30740/TCP,9101:32698/TCP,8080:30950/TCP   3m41s
fullnode-78d29195        ClusterIP   10.64.10.80   <none>        8080/TCP                                                      61s

$ kg pvc                                    
NAME                                   STATUS   VOLUME                                     CAPACITY   ACCESS MODES   STORAGECLASS   AGE
aptos-data-fullnode-78d29195-0         Bound    pvc-03c92421-4034-48c4-8bf3-5696462078d0   2Ti        RWO            standard-rwo   70s
aptos-node-0-validator-eforge116       Bound    pvc-093cb40e-dbe3-4ab0-98a0-a9f80cc3a776   2Ti        RWO            ssd            3m50s
fn-aptos-node-0-fullnode-eforge116-0   Bound    pvc-3929e63d-daf1-4a96-903d-b70fd20f4b78   2Ti        RWO            ssd            3m49s
```

* [Grafana dashboard (auto-refresh)](https://aptoslabs.grafana.net/d/overview/overview?orgId=1&refresh=10s&var-Datasource=VictoriaMetrics%20Global%20%28Non-mainnet%29&var-BigQuery=Google%20BigQuery&var-namespace=forge-rustielin-1686186329&var-metrics_source=All&var-chain_name=forge-0&refresh=10s&from=now-15m&to=now)
* [Humio Logs](https://cloud.us.humio.com/k8s/search?query=%24forgeLogs%28validator_instance%3D%2A%29+%7C%0A++++%22k8s.namespace%22+%3D+%22forge-rustielin-1686186329%22+%2F%2F+filters+on+namespace+which+contains+validator+logs%0A+++OR++%2F%2F+remove+either+side+of+the+OR+operator+to+only+display+validator+or+forge-runner+logs%0A++++%22k8s.labels.forge-namespace%22+%3D+%22forge-rustielin-1686186329%22+%2F%2F+filters+on+specific+forge-runner+pod+in+default+namespace%0A&widgetType=list-view&columns=%5B%7B%22type%22%3A+%22field%22%2C+%22fieldName%22%3A+%22%40timestamp%22%2C+%22format%22%3A+%22timestamp%22%2C+%22width%22%3A+180%7D%2C+%7B%22type%22%3A+%22field%22%2C+%22fieldName%22%3A+%22level%22%2C+%22format%22%3A+%22text%22%2C+%22width%22%3A+54%7D%2C+%7B%22type%22%3A+%22link%22%2C+%22openInNewBrowserTab%22%3A+%22%2A%2A%2A%22%2C+%22style%22%3A+%22button%22%2C+%22hrefTemplate%22%3A+%22https%3A%2F%2Fgithub.com%2Faptos-labs%2Faptos-core%2Fpull%2F%7B%7Bfields%5B%5C%22github_pr%5C%22%5D%7D%7D%22%2C+%22textTemplate%22%3A+%22%7B%7Bfields%5B%5C%22github_pr%5C%22%5D%7D%7D%22%2C+%22header%22%3A+%22Forge+PR%22%2C+%22width%22%3A+79%7D%2C+%7B%22type%22%3A+%22field%22%2C+%22fieldName%22%3A+%22k8s.namespace%22%2C+%22format%22%3A+%22text%22%2C+%22width%22%3A+104%7D%2C+%7B%22type%22%3A+%22field%22%2C+%22fieldName%22%3A+%22k8s.pod_name%22%2C+%22format%22%3A+%22text%22%2C+%22width%22%3A+126%7D%2C+%7B%22type%22%3A+%22field%22%2C+%22fieldName%22%3A+%22k8s.container_name%22%2C+%22format%22%3A+%22text%22%2C+%22width%22%3A+85%7D%2C+%7B%22type%22%3A+%22field%22%2C+%22fieldName%22%3A+%22message%22%2C+%22format%22%3A+%22text%22%7D%5D&newestAtBottom=%2A%2A%2A&showOnlyFirstLine=false&live=true&start=30m)
* See in the API dashboard that the traffic is indeed being served by the new PFN: [Grafana dashboard for API](https://aptoslabs.grafana.net/d/api/api?from=now-15m&to=now&var-Datasource=VictoriaMetrics%20Global%20%28Non-mainnet%29&var-BigQuery=Google%20BigQuery&var-metrics_source=All&var-chain_name=forge-0&var-cluster=All&var-namespace=forge-rustielin-1686186329&var-kubernetes_pod_name=fullnode-78d29195-0&orgId=1&refresh=30s&var-endpoint=All&var-role=All&var-interval=1m&var-status=All&var-operation_id=All)

<!-- Please provide us with clear details for verifying that your changes work. -->

perryjrandall commented on testsuite/forge-test-runner-template.yaml:
```diff
@@ -11,7 +11,7 @@ spec:
   containers:
   - name: main
-    image: {FORGE_IMAGE_REPO}:{FORGE_IMAGE_TAG}
+    image: {IMAGE_REPO}/forge:{FORGE_IMAGE_TAG}
```
why are we doing this concat here?

Is possible I'd like to move as much logic from templates out

So in that case it would just be a single {IMAGE_LOCATION} and you can do the concat in rust

perryjrandall replied:
The problem here is that this makes our helm templates contain more logic and it makes it more ingrained

Ultimately we wanna move away from relying on the template implementation and therefore i think it makes more sense to move that kind of logic higher level

so the semantic variables passed in stand on their own, which makes it easier to swap out template implementation, because the template is stupid (think of what variables we might wanna pass to pangu)

rustielin replied:
tru ok ya done

perryjrandall commented on testsuite/forge/src/backend/k8s/mock_kube_api.rs:
```diff
@@ -0,0 +1,201 @@
+// Copyright © Aptos Foundation
```
You might consider moving this into kube_api with a #[cfg(test)] that way the testing mocks are tightly coupled with their sources

rustielin replied:
hrm yea good point!

perryjrandall commented on testsuite/forge/src/backend/k8s/fullnode.rs:
```diff
@@ -0,0 +1,738 @@
+                    ..validator_stateful_set
+                        .spec
+                        .as_ref()
+                        .unwrap()
```
I dont think we should unwrap here, we can raise the error upwards, but unwrapping will panic in a cryptic way

perryjrandall commented on testsuite/forge/src/backend/k8s/fullnode.rs:
```diff
@@ -0,0 +1,738 @@
+                        .spec
+                        .as_ref()
+                        .unwrap()
+                        .containers[0]
```
What if containers is empty? Can we use first and map error?

perryjrandall commented on testsuite/forge/src/backend/k8s/fullnode.rs:
```diff
@@ -0,0 +1,738 @@
+                ..validator_stateful_set
+                    .spec
+                    .as_ref()
+                    .unwrap()
```
same comment

perryjrandall commented on testsuite/forge/src/backend/k8s/mod.rs:
```diff
@@ -178,6 +186,8 @@ impl Factory for K8sFactory {
+            self.use_port_forward,
         )
         .await
         .unwrap();
```
Can you remove this unwrap while you're here?

perryjrandall reviewed (approved):
Yuge PR

This looks great generally, amazing job testing 🙏 

Left comments, mostly on style

The two salient themes from comments

1) Lets try and avoid unwrap / panic when possible because it makes cryptic errors and doesnt allow the caller to make the decision when to panic
2) For large nested structures, we might wanna consider a builder pattern if we're going to be building al ot of these things, it can make the code a lot easier to read, you could even make a dynamic resource builder that can coerce to different resources as well using TValue since they're all serializable

But overall amazing, this unlocks a big need for forge

perryjrandall commented on testsuite/forge/src/backend/k8s/fullnode.rs:
```diff
@@ -0,0 +1,738 @@
+    let mut fullnode_stateful_set = StatefulSet::default();
+    fullnode_stateful_set.metadata.name = Some(fullnode_name.clone());
+    fullnode_stateful_set.metadata.labels = Some(labels_map.clone());
+    fullnode_stateful_set.spec = Some(StatefulSetSpec {
```
Yeah I'm not suggesting we use stackable operator, just that you can use that same kind of pattern

rustielin replied:
ah I see. yea it's a bit hard to read rn, lemme see what i can do...

rustielin commented on config/src/config/node_config.rs:
```diff
@@ -228,9 +228,21 @@ fn parse_serialized_node_config(serialized_config: &str, caller: &'static str) -
+        node_config,
+        override_node_config,
+    )
+    .map_err(|e| anyhow::anyhow!("Unable to merge default config with override. Error: {}", e))
```
truee

sitalkedia commented on testsuite/forge/src/backend/k8s/fullnode.rs:
```diff
@@ -0,0 +1,738 @@
+const APTOS_DATA_VOLUME_PATH: &str = "/opt/aptos/data";
+
+/// Get the image repo from the validator. The fullnode will use the same image repo, as they both use the "validator" image
+fn get_image_repo_from_validator(validator_stateful_set: &StatefulSet) -> Result<String> {
```
Can you reuse the `get_stateful_set_image` function instead?

rustielin replied:
ah you're right! I forgot about that function... 🤦

sitalkedia reviewed (approved):
LGTM % minor nits. Thank you for adding this long awaited feature to Forge. 🙌

rustielin commented on testsuite/forge/src/backend/k8s/fullnode.rs:
```diff
@@ -0,0 +1,738 @@
+                        .containers[0]
+                        .clone()
+                }],
+                volumes: Some(vec![
```
yea good idea

rustielin commented:
Thanks for the reviews all! @perryjrandall @sitalkedia 
I've gone ahead and removed all non-test uses of unwrap, and also cleaned up the way the main fullnode StatefulSet is created, to be much more readable.

rustielin commented:
Canary looks good! https://github.com/aptos-labs/aptos-core/actions/runs/5216952842/jobs/9416424905
